numba/cuda/cudadrv/devicearray.py

```diff
@@ -354,15 +354,38 @@ def view(self, dtype):
         copy of the data.
         """
         dtype = np.dtype(dtype)
-        if dtype.itemsize != self.dtype.itemsize:
-            raise TypeError("new dtype itemsize doesn't match")
+
+        if self.dtype.itemsize != dtype.itemsize and not self.is_c_contiguous():
+            raise ValueError(
+                "To change to a dtype of a different size,"
+                " the array must be C-contiguous"
+            )
+
+        shape = list(self.shape)
+        shape[-1], rem = divmod(
+            shape[-1] * self.dtype.itemsize,
+            dtype.itemsize
+        )
+        shape = tuple(shape)
+
+        if rem != 0:
+            raise ValueError(
+                "When changing to a larger dtype,"
+                " its size must be a divisor of the total size in bytes"
+                " of the last axis of the array."
+            )
+
+        strides = list(self.strides)
+        strides[-1] = dtype.itemsize
```

Comment:
This modification of strides is incorrect when `self.strides` is not equal to dtype (e.g. for a view like `arr[:, ::2]`)

Reply:
Yep you're right. In that case the only thing that is possible is to preserve `itemsize` when changing type. So the strides should remain the same. Thanks for catching that. 🙂